### PR TITLE
fix(task_executor): fix Langfuse flush/shutdown deadlock that freezes document parsing

### PR DESCRIPTION
**Pull request by ondery**
## Summary

Parsing large documents (e.g. a 1500-page PDF) eventually **freezes the whole task executor**: a few page tasks finish, then every document stays stuck and newly queued documents never start. Stopping and re-parsing does not help and the UI shows an ever-growing `N tasks are ahead in the queue...` (e.g. `127 -> 254 ...`).

This PR fixes the actual freeze (a Langfuse flush/shutdown deadlock) and makes the queue figure honest. It does **not** change how documents are split, so large PDFs keep parsing in full.

## 1. Root cause of the freeze (the real bug)

`LLM4Tenant.close()` runs once per task, synchronously, on the task executor's asyncio event-loop thread (via `LLMBundle.__exit__`). It called:

```python
self.langfuse.flush()
if hasattr(self.langfuse, 'shutdown'):
    self.langfuse.shutdown()
```

Both are unsafe there:

- `flush()` blocks on an unbounded `queue.join()` in the underlying OpenTelemetry span processor and never returns if the exporter cannot drain.
- `shutdown()` permanently tears down the **process-wide** Langfuse / OpenTelemetry tracer provider that every `LLMBundle` shares. After the first task shuts it down, every subsequent `flush()` blocks forever.

Because this runs on the event loop, one stuck `flush()` freezes the entire executor. Captured live with `py-spy` on a stuck executor — the `MainThread` (event loop) was parked here while all other threads were idle:

```
join (queue.py:105)
flush (.../opentelemetry .../resource_manager.py:445)
flush (langfuse/.../client.py:2116)
close (api/db/services/tenant_llm_service.py:545)
close (api/db/services/llm_service.py:49)
__exit__ (api/db/services/llm_service.py:57)
handle (task_handler.py:144)
...
handle_task (task_executor.py:1713)
```

**Fix:** drop the Langfuse reference in `close()` without flushing/shutting down the shared client. Langfuse exports spans from its own background processor and flushes at process exit, so traces are still delivered.

## 2. Honest, self-healing queue length

The `N tasks are ahead in the queue...` text came straight from the Redis stream consumer-group `lag`, which counts every undelivered message — including messages whose tasks were already cancelled/stopped, and a document's own not-yet-started page tasks. So a single big PDF reported its own 127 page tasks as "ahead", and the number could appear to grow on every cancel+restart because stale messages are never removed from the stream.

- `get_queue_length()` now caps the raw Redis `lag` by the number of tasks genuinely still waiting in the database (`Task.progress == 0` on an actively-running document), with a short monotonic TTL cache to avoid a `COUNT` per document each progress cycle. Falls back to the raw `lag` when the DB count is unavailable.
- `_sync_progress` subtracts the document's own queued tasks for the matching priority queue, so the figure reflects how many *other* tasks are ahead (addresses CodeRabbit's priority-scoping note from #16499).
- The stale `N tasks are ahead in the queue...` message is cleared when a parse is stopped/cancelled (single-doc run, bulk stop, per-task stop).

## Supersedes

Replaces #16499 (queue-length-only fix); this PR contains those changes plus the deadlock fix as a single change set.

### CODE REVIEW

**Comment by ondery** [comment]
Addressed the CodeRabbit review feedback in 3be9a044d:

- **Priority-specific DB cap (major):** `get_pending_task_count(priority)` now filters by `Task.priority` and caches per priority; `get_queue_length` passes its priority through. A stale lag on one priority queue no longer blocks self-healing to 0 when that priority has no real pending work.
- **Ruff B009** (`tenant_llm_service.close`): replaced `callable(getattr(self.mdl, "close"))` with `callable(getattr(self.mdl, "close", None))`.
- **Ruff S110/BLE001** (test helper): the cv2 stub now catches `ImportError` and logs instead of a bare `except Exception: pass`; tests updated for the per-priority cache (plus a new per-priority cache test).
- **Logging for new flows:** added `logging.debug` at the cancel/stop `progress_msg` cleanup points (doc_id / task_id).

File: api/db/services/document_service.py [hunk 1]
```diff
@@ -1102,8 +1114,70 @@ def new_task():
...
+    try:
+        query = (
+            Task.select(fn.COUNT(Task.id))
+            .join(Document, on=(Task.doc_id == Document.id))
+            .where(
+                (Task.progress == 0)
+                & (Document.run == TaskStatus.RUNNING.value)
+                & (Document.progress >= 0)
+                & (Document.progress < 1)
+            )
```
[the remark this answers is not in the document]
Comment (ondery) [reply api/db/services/document_service.py:1218]: Fixed in 931d5429c. `get_pending_task_count` no longer requires `run == RUNNING` / `progress < 1`; it now counts progress==0 tasks unless the document is cancelled (`run == CANCEL`), failed (`progress < 0`) or soft-deleted (`status != VALID`). This keeps `begin2parse(keep_progress=True)` graphrag/raptor tasks in the count so the cap no longer wrongly drops to 0 while lag is non-zero.

File: api/db/services/document_service.py [hunk 2]
```diff
@@ -1102,8 +1114,70 @@ def new_task():
...
+    that are genuinely still waiting in the database, which self-heals the
+    moment work is cancelled or completes.
+    """
     group_info = REDIS_CONN.queue_info(settings.get_svr_queue_name(priority, suffix), SVR_CONSUMER_GROUP_NAME)
-    if not group_info:
-        return 0
-    return int(group_info.get("lag", 0) or 0)
+    lag = int(group_info.get("lag", 0) or 0) if group_info else 0
+
+    pending = get_pending_task_count(priority)
```
[the remark this answers is not in the document]
Comment (ondery) [reply api/db/services/document_service.py:1252]: Fixed in 931d5429c: `get_queue_length` now short-circuits and returns 0 when `lag <= 0`, so the DB COUNT/JOIN is skipped entirely in the common empty-queue case during the progress-sync loop.

File: test/unit_test/api/db/services/test_get_queue_length.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    message="pkg_resources is deprecated as an API.*",
+    category=UserWarning,
+)
+
+
+def _install_cv2_stub_if_unavailable():
+    try:
+        import cv2  # noqa: F401
+        return
+    except ImportError as exc:
```
[the remark this answers is not in the document]
Comment (ondery) [reply test/unit_test/api/db/services/test_get_queue_length.py]: Fixed in 931d5429c: the cv2 stub helper now catches `(ImportError, OSError)` so missing shared libraries no longer break test module import, while still avoiding a fully blind `except`.

File: api/db/services/document_service.py [hunk 4]
```diff
@@ -1102,8 +1114,79 @@ def new_task():
...
+        return cached["value"]
+    try:
+        query = (
+            Task.select(fn.COUNT(Task.id))
+            .join(Document, on=(Task.doc_id == Document.id))
+            .where(
+                (Task.progress == 0)
+                & (Document.run != TaskStatus.CANCEL.value)
+                & (Document.progress >= 0)
+                & (Document.status == StatusEnum.VALID.value)
```
[the remark this answers is not in the document]
Comment (ondery) [reply api/db/services/document_service.py]: Fixed in 82e4892bb: the pending-count query now uses `(Document.run.is_null(True)) | (Document.run != TaskStatus.CANCEL.value)`, matching the rest of DocumentService so legacy/raw-inserted documents with NULL `run` are counted as non-cancelled and no longer undercount the queue length.

**Review by Lynn-Inf — CHANGES_REQUESTED** [review]
Thanks for the PR. A few comments:

1. Use `_session_id` and `_user_id` as defaults in `agent/canvas.py:446-447` instead of null to keep `llm_request_context` clean.

2. In `task_api.py:92`, I noticed you're setting `progress_msg` to an empty string. Could you explain the reasoning behind this change?

3. Changes to `LLM4Tenant` may affect `task_executor` — I'll ask @xugangqiang to evaluate that part.

**Comment by ondery** [comment]
> Thanks for the PR. A few comments:
> 
> 1. Use `_session_id` and `_user_id` as defaults in `agent/canvas.py:446-447` instead of null to keep `llm_request_context` clean.
> 2. In `task_api.py:92`, I noticed you're setting `progress_msg` to an empty string. Could you explain the reasoning behind this change?
> 3. Changes to `LLM4Tenant` may affect `task_executor` — I'll ask @xugangqiang to evaluate that part.

### 1. Issue

I had already opened **PR #16429** for this change. However, when creating **PR #16502**, I accidentally based it on the previous branch, so the PRs became stacked. I explained the background and the reasoning in more detail in **#16429**.

In short, **Langfuse** uses `user_id` and `session_id` for tracing and user-level analytics. **OpenRouter** also surfaces these fields in its Logs page, allowing requests to be correlated with individual users and sessions. I couldn't find any other reliable way to propagate this information through the current architecture. Without these identifiers, it becomes nearly impossible to distinguish requests in a multi-user (multi-tenant) environment.

---

### 2. Issue

This change is related to the **Stop Parsing** behavior.

When a parsing task became stuck and the user clicked **Stop Parsing**, the details page continued to display the message:

> **"N tasks are ahead in the queue..."**

Even though the parsing process had already been terminated, this message remained visible, which was misleading because it no longer reflected the actual state of the task.

I initially created **PR #16499** to address only this UI/message issue. Later, I concluded that it made more sense to solve both the parsing lock-up issue and the misleading status message together. Therefore, I closed **#16499** and consolidated everything into **PR #16502**.

---

### 3. Issue

Regarding **"Changes to LLM4Tenant may affect task_executor"**, that's a reasonable concern.

However, I have been running this version for several days in my own environment without encountering any issues related to `task_executor`. Of course, there may still be edge cases that I haven't covered, but based on my current testing, I haven't observed any regressions caused by these changes.

**Comment by ondery** [comment]
I'd also like to provide some additional context regarding **issue 1**.

I'm using RAGFlow as the AI backend for my own **multi-tenant** platform. Ideally, I would create a separate RAGFlow user for every tenant because it provides better isolation. However, that approach is not practical for my use case.

Each user in my platform needs access to the same set of agents, search configurations, and shared datasets, while also having their own private datasets. At the moment, there isn't a built-in mechanism to define reusable **agent templates** or automatically provision a new RAGFlow user with a predefined configuration.

In my scenario, whenever a new user registers, my backend would have to:

* Create a new RAGFlow user.
* Recreate all required agents from an existing template.
* Replicate the necessary configuration.
* Invite the user to every shared dataset and wait for those invitations to be accepted.

Besides adding considerable operational complexity, the dataset invitation flow is not suitable for an automated SaaS onboarding process, since I cannot expect every newly registered user to manually accept invitations.

Because of these limitations, I currently manage all requests through a single RAGFlow account and distinguish users at the application layer by attaching custom metadata to every request.

For this reason, propagating identifiers such as `user_id` and `session_id` to integrations like Langfuse and OpenRouter is essential. These identifiers allow me to correlate requests, traces, logs, and usage with the actual end user, which is critical in a multi-tenant environment.

**Comment by Lynn-Inf** [comment]
Thanks for the context.

1. For multi‑tenant scenarios, consider using the **Team** feature — invite users as members to avoid re‑creating agents/configs, and keep private datasets isolated.

2. For the misleading queue message, a better fix might be to append a "cancelled" log to `progress_msg` instead of resetting it. This preserves the history and clearly indicates the task was stopped.

**Review by wangq8 — COMMENTED** [review]
## Review of PR #16502

**Overall: Looks good.** The PR is well-structured, addresses a real production deadlock with solid root-cause analysis (confirmed via `py-spy`), and includes comprehensive unit tests. 511 additions / 29 deletions across 10 files is a reasonable size for this fix.

---

### 1. Root cause fix (Langfuse flush/shutdown deadlock) — Excellent

The diagnosis is spot-on: `flush()` blocks on an unbounded `queue.join()` inside the OpenTelemetry span processor, and `shutdown()` permanently tears down the process-wide tracer provider. Doing either on the event-loop thread of the task executor is the kind of bug that would be incredibly painful to debug without `py-spy` — well caught.

The fix (just dropping the reference without flushing/shutting down) is correct. Langfuse exports from its own background processor and flushes at process exit, so traces are still delivered.

**Bonus:** The docstring in `close()` is thorough, explaining not just *what* changed but *why* — that is exactly the level of documentation this kind of fix deserves. The improved logging (warning with `exc_info` instead of silent `pass`) is also a nice upgrade.

---

### 2. Queue-length capping — Good

Capping Redis stream lag (`group_info["lag"]`) with the genuine DB pending task count is the right approach. A few observations:

- **TTL cache (3s)** — reasonable balance between accuracy and avoiding a COUNT/JOIN per document per cycle.
- **Per-priority cache keying** — correctly prevents one priority queue with stale lag from polluting another.
- **`_sync_progress` subtraction** — excluding the document's own queued tasks from "ahead" is the right semantic fix. The `max(0, ...)` guard handles the edge case correctly.

**Minor observation:** `_PENDING_TASK_COUNT_CACHE` is module-level mutable state accessed from potentially multiple task-executor threads. For a TTL-based cache this doesn't need strict locking (stale reads self-heal within 3s), but worth noting if anyone ever extends the cache lifespan.

---

### 3. Progress message cleanup — Good

Clearing `progress_msg` on cancel/stop paths is the correct UI fix. The logging added (per CodeRabbit suggestion) at each cleanup point makes the flow traceable.

---

### 4. LLM request context propagation — Good

New `common/llm_request_context.py` is clean:

- ContextVar pattern with proper `set`/`reset` lifecycle and stale-token fallback.
- Session ID preferred over user ID, 128-char truncation, debug logging avoids raw identifiers.
- The `"user" not in completion_args` check correctly preserves caller-supplied `user` (including empty string to suppress forwarding).

The `canvas.py` change that copies contextvars before `run_in_executor` is necessary because context variables don't propagate across threads — glad to see this was handled rather than just setting the ContextVar and hoping for the best.

---

### 5. Test coverage — Good

- **`test_get_queue_length.py`** — 6 test cases covering: DB cap, self-heal to zero, lag-as-bound, DB fallback, missing group info, null lag. Good coverage of edge cases.
- **`test_llm_request_context.py`** — 9 test cases covering: no context, session preference, user fallback, empty IDs, truncation, reset lifecycle, stale token resilience, and `completion_args` `user`-field precedence. Thorough.

Both test files include proper fixture isolation and the `cv2` stub is well-handled.

---

### 6. Minor nits (non-blocking)

- **Module namespace pollution approach in `get_queue_length`:** The module-level `_PENDING_TASK_COUNT_CACHE` and TTL constant work fine, but if this pattern grows, consider wrapping in a small class with classmethods for self-contained cache management.
- **`canvas.py` context variable copy:** The old code `lambda: ctx.run(partial(...))` was unnecessarily nested; the new version using `partial(call_ctx.run, bound_call)` is cleaner. Good cleanup.

---

### Summary

| Area | Verdict |
|---|---|
| Deadlock fix | Correct and well-explained |
| Queue length honesty | Well-reasoned with appropriate caching |
| Progress msg cleanup | Sound |
| LLM context propagation | Clean implementation |
| Tests | Thorough |
| Docs/comments | Excellent |

This is a solid PR that fixes a real, painful bug. The only open concern from the RAGFlow team (Lynn-Inf's point about `LLM4Tenant` changes potentially affecting `task_executor`) should get a second pair of eyes from xugangqiang, but the change itself is logically sound.

**Approve from a code-quality standpoint.**

File: agent/canvas.py [hunk 5]
```diff
@@ -438,6 +439,13 @@ async def run(self, **kwargs):
             _lf_attrs["session_id"] = str(_session_id)[:200]
         sink_token = token_usage_sink.set(self._run_token_usage)
         attrs_token = langfuse_run_attrs.set(_lf_attrs)
+        # Forward the originating session/user to upstream LLM providers (as the
+        # OpenAI `user` field) for the duration of this run, and reset afterwards so
+        # the value never leaks to later calls in the same task.
+        _req_ctx_token = set_llm_request_context(
+            session_id=kwargs.get("session_id"),
```
Comment (Lynn-Inf) [inline agent/canvas.py]: Better to use `_session_id` and `_user_id` rather than get from kwargs again.

File: api/apps/restful_apis/task_api.py [hunk 6]
```diff
@@ -89,7 +89,8 @@ async def _cancel_task(task_id):
         if doc_id and doc_id not in (CANVAS_DEBUG_DOC_ID, GRAPH_RAPTOR_FAKE_DOC_ID):
             _, doc = DocumentService.get_by_id(doc_id)
             if doc and str(doc.run) in (TaskStatus.RUNNING.value, TaskStatus.SCHEDULE.value):
-                DocumentService.update_by_id(doc_id, {"run": TaskStatus.CANCEL.value, "progress": 0})
+                DocumentService.update_by_id(doc_id, {"run": TaskStatus.CANCEL.value, "progress": 0, "progress_msg": ""})
```
Comment (Lynn-Inf) [inline api/apps/restful_apis/task_api.py]: Either keep `progress_msg` unchanged or append a "cancelled" marker.

Comment (ondery) [reply api/apps/restful_apis/task_api.py]: Done in 998e5e971. Instead of resetting `progress_msg` to an empty string, the cancel/stop paths (`task_api.py` and both cancel points in `document_api.py`) now append a `"\nHH:MM:SS Task stopped by user."` marker to the existing `progress_msg`. This preserves the parse history and clearly indicates the task was stopped.

File: agent/canvas.py (hunk 5 above)
Comment (ondery) [reply agent/canvas.py]: Done in 998e5e971. `set_llm_request_context` now reuses the already-derived `_session_id`/`_user_id` instead of reading `kwargs` again, so the Langfuse attributes and the upstream `user` forwarding stay consistent (and `_session_id` keeps its `self._id` fallback).
